### ISSUE TITLE
"Processing at Scale" clarifications on creating additional runners...

### ISSUE THREAD

**Issue by EliW** [Prospress/action-scheduler#874]
Hey team ... Thanks for earlier help/clarification on docs.  And I've got one more.   We are using Action Scheduler now (yay) in production to help offload some background async tasks, and it's working great.  Except for the fact that I forgot to apply the 'run at scale' things at first, and so tasks were happening in a single queue.   Ooops.

So I went to start adding the code per: https://actionscheduler.org/perf/ to enable multiple queues, and ran into a bit of a conundrum, and/or I don't grok something.  

It's important for us to increase the number of concurrent running tasks, because part of why we are offloading background processing pieces here is that we have lots of them that are working with a slow 3rd party API, and might take 20-40 seconds for the task to complete.  Perfect to fire off async.  But when you take an admin action to say, update 400 records, and each one can take that long.   Doing it one-at-a-time is painful.

So changing the runtime, batch size, and current-allowed is all good.   But then I realized per the docs that just changing the concurrent setting isn't enough.

We are hosting on WPengine, and we have 'Alternate Cron' enabled, which means that every minute WPengine triggers our cron processes to fire.   But per the docs, that does mean that as long as our processes can't take longer than a minute anyway, that we'll only ever have 1 concurrent queue running.

I went to look at the code that was suggested to add to actually create multiple concurrent queues ... but realized that it appears ... brute force?   If I'm understanding the sample code correctly, it will every minute, fire off 5 additional requests back to the server via admin-ajax.  Regardless of how many items are in the queue.   Which seems rather harsh to the server, especially if most of the time (with our current use of this) that there won't be anything in the queue.  Yet every minute we'll cause 5 extra hits to the server to try to run more queues, to handle ... nothing?

I started digging through the code, to find a way to only make it spawn more concurrent requests *if* there were items in the queue.   But before I went too deep there I wanted to pop on here and ask, because if there was a zero-impact way of doing that, I'd have expected to see that in the base example, which made me go 'hrmmmm'.

PS.  Additionally I assume that there is no way by default to set 'weights' to queue items?   Trying to balance the batch size and realizing I've got a catch-22.  Because in general all the tasks we will be putting into the queue are ways to offload time-intensive items.   But those can vary.  One type might be ~1sec, another might be ~5sec, and another that ~40sec....    I realized that it would be good to have a way to not just specify "batch size" but a "weighted batch size".  Where we basically tell it "take ~50 seconds of stuff", and somehow when scheduling the actions give it a 'approx length / weight'.   ---  Right now for example I'm tempted for our use case to drop the batch size to '2' so that more concurrent queues can grab more longer running things.   Assuming my 'smart' concurrent queue concerns above can be answered ...

P.PS.   I would assume that another way of making more concurrent requests would be to hit the wp-cron.php endpoint more frequently than once-a-minute?   (I'd have to check with WPengine as to whether they could do that for us, or if we'd have to do that ourself ... unless wp-cron might ignore more frequent requests itself anyway)

**Thanks again so much for making this awesome tool for WP folks.  Looking forward to hearing back!**

**Comment by barryhughes** [comment]
Hi @EliW,

> If I'm understanding the sample code correctly, it will every minute, fire off 5 additional requests back to the server via admin-ajax. Regardless of how many items are in the queue. Which seems rather harsh to the server, especially if most of the time (with our current use of this) that there won't be anything in the queue. Yet every minute we'll cause 5 extra hits to the server to try to run more queues, to handle ... nothing?

I think you've understood correctly, here. I understand your apprehension about this, though:

- Those queue runner processes should stand down and exit very quickly if there is nothing left to be processed.
- It would be possible to make a further modification, which examines the number of waiting actions (and checks how many active claims exist) *before* deciding how many additional runners to spawn.

> PS. Additionally I assume that there is no way by default to set 'weights' to queue items?

Not exactly; but in the long-term we wish to introduce action priorities. 

Additionally, especially if you have access to WP CLI, you can create queue runners that focus on specific action groups, *or* specific hooks, or indeed some combination of those things—and you can specify different batch sizes for each. Again, as long as you lift the limit on how many concurrent queue runners are allowed, you can run these side-by-side (but I'm not sure if WP CLI, or system crontab, is accessible in your environment).

> P.PS. I would assume that another way of making more concurrent requests would be to hit the wp-cron.php endpoint more frequently than once-a-minute?

I'm not sure that would work, at least not with an 'out of the box' configuration, because Action Scheduler registers an event that should run every minute—and WP Cron will (try to) respect that and generally won't run events ahead-of-time.

**Comment by EliW** [comment]
> Hi @EliW 

Hey @barryhughes - Thanks for the response!

> I think you've understood correctly, here. I understand your apprehension about this, though:
>
> Those queue runner processes should stand down and exit very quickly if there is nothing left to be processed.

True.  However it would still mean that 'every minute', we'd issue (5 in the example) additional requests back to our server.   Meaning that we'd still be adding 'connection overhead' if these weren't needed at the time.   Perhaps worrying about this is pre-mature optimization.  But since our current environment isn't 100% under our control, I just worry about things that might end up slowing things down.

> It would be possible to make a further modification, which examines the number of waiting actions (and checks how many active claims exist) before deciding how many additional runners to spawn.

Do you happen to have any sample code for how to do this?  There aren't any public functions for this.  I started digging into the classes and looks like I'd need to call `ActionScheduler_Store::query` ... but looking at it, it appears perhaps to be a maybe expensive operation.

>> PS. Additionally I assume that there is no way by default to set 'weights' to queue items?

> Not exactly; but in the long-term we wish to introduce action priorities.

That sounds cool.  I'll throw my hat in the ring for 'action weights' as well then ('actions sizes' if you don't like weights?   or 'scale'?).   It would be nice to be able to state when it's doing "get batch size 25" ... for it to see various weights and take those into account.   So we could put some items in at weight 1 (default).   And others at weight 5, or 20 ... to state that this one we expect to take 20x as much effort as something with weight 1.   And so that batch size 25 might grab 25 'weight 1' items, or a weight-20 + weight+5, or any combination that adds up to 25 or less.

> (but I'm not sure if WP CLI, or system crontab, is accessible in your environment).

We don't.  Honestly if we had access to the system crontab we likely wouldn't even be looking at using ActionScheduler. 

> > P.PS. I would assume that another way of making more concurrent requests would be to hit the wp-cron.php endpoint more frequently than once-a-minute?

> I'm not sure that would work, at least not with an 'out of the box' configuration, because Action Scheduler registers an event that should run every minute—and WP Cron will (try to) respect that and generally won't run events ahead-of-time.

I may do some tests then.  Because the idea was that we'd change Action Scheduler's event to run every 5 seconds for example.  But then tell it to allow 10 concurrent.  And theoretically WPcron would be cool w/ that.

**Comment by barryhughes** [comment]
> Do you happen to have any sample code for how to do this?

I'm afraid we don't have any sample code to hand (regarding a tweak to the [`eg_request_additional_runners()` sample code](https://actionscheduler.org/perf/#increasing-initialisation-rate-of-runners), that is) but yes: I imagine you would probably want to take advantage of a custom query, here—though, you could consider hand-crafting your own queries, if the performance of Action Scheduler's query methods is a concern.

> I'll throw my hat in the ring for 'action weights' as well then

I see now you are picturing something a little different to priorities (as we had envisioned them). That's a really interesting idea, though, and I can see the potential value there.

> We don't. Honestly if we had access to the system crontab we likely wouldn't even be looking at using ActionScheduler.

Since it's not available this is rather moot, but I'll comment anyway (if only for the benefit of others coming across this conversation) that the combination of a system scheduler (ie, crontab) and Action Scheduler is very appealing in many cases, and offers a lot of flexibility that is difficult to accomplish with crontab alone.

> Because the idea was that we'd change Action Scheduler's event to run every 5 seconds for example. But then tell it to allow 10 concurrent. And theoretically WPcron would be cool w/ that.

Apologies, as I misunderstood you first time around. Yes, if you register then assign a new schedule ('run every 5 seconds'), and can trigger wp-cron.php with the same or higher frequency, this is likely a viable approach.

**Comment by EliW** [comment]
> Since it's not available this is rather moot, but I'll comment anyway (if only for the benefit of others coming across this conversation) that the combination of a system scheduler (ie, crontab) and Action Scheduler is very appealing in many cases, and offers a lot of flexibility that is difficult to accomplish with crontab alone.

OH YES!  My apologies as I didn't mean for that to come across negative in any way towards Action Scheduler.  Just in our particular use case part of the appeal of Action Scheduler is that it works "even if you don't have Cron".  Part of the same reason WPcron exists as well.    I found Action Scheduler because of that.  But currently we are using it for the async/ondemand tasking, very much in the same way you might use RabbitMQ or other such systems to fire off tasks to happen 'when it gets a chance' ;)    

Awesome tool you've made here (and hence the conversations).

> Apologies, as I misunderstood you first time around. Yes, if you register then assign a new schedule ('run every 5 seconds'), and can trigger wp-cron.php with the same or higher frequency, this is likely a viable approach.

Cool.  I started doing some testing locally on that, but oddly wasn't able to get multiple queues to run.  (I assume if it works I should see multiple task in-progress at the same time, right?) Need to do some more to find out why the wp-cron method wasn't triggering that.  I set a '5 second' time for Action Scheduler and I can see the cron gettting triggered/reset.  But no bonus queues appear to happen.  I must have messed something up :)   

(Or there's something 'deeper' going on)

Shoot, found it.  That's not going to work #sigh.   `WP_CRON_LOCK_TIMEOUT` is not our friend.  so wp-cron itself won't let another cron spawn do anything as long as another cron instance is running.   And it uses that const to figure that out. And it defaults to 1 minute.   Now, I COULD set `WP_CRON_LOCK_TIMEOUT` to a lower number. But not sure that's a good idea.  Since that can/will cause overlapping cron (which I realize was EXACTLY what I was trying to do).   But if WP core is trying to not have them overlap.   Then there must be a solid reason for that besides "because" and some side-effects might occur.   So yeah, I guess I'm back to doing the original way from the docs.  Since that doesn't actually use cron at all.

... THOUGH ...

It has hit me that another solution here (should I wish), would be to not triggers those admin-ajax requests in the `action_scheduler_run_queue` cron event.   But instead if I was going to be sending a wp-cron every 5 seconds anyway.  I could instead just hit that queue endpoint directly instead perhaps.

**Comment by barryhughes** [comment]
Right, I forgot about the lock; and indeed I think WP Cron stores its list of events in a single serialized option (as opposed to a table of individual rows), which might be a big part of why they want to avoid the equivalent of concurrent queue runners in their specific case.

Simply playing with ideas here, this may not be the best solution, but also riffing on your concept of actions having different weights (and the fact you have some that interact with potentially slow remote services) ... what about a pattern along these lines:

```php
as_schedule_single_action( 
	$some_time, 
	'run_heavy_weight_action', 
	[ 'callback' => 'interact_with_slow_service' ] 
);

function run_heavy_weight_action( callable $callback ) {
	possibly_spawn_additional_queue_runner();
	call_user_func( $callback );
}

function interact_with_slow_service() {
	/* Does whatever this function does. */
}
```

Function `possibly_spawn_additional_queue_runner()` would, as the name suggests, possibly spawn another queue runner by async HTTP request, using a lock and/or counter to limit how many are created in this way.
